### PR TITLE
SQL project and filter operators

### LINKED ISSUE #16893: SQL project and filter operators
We need to implement project and filter operators for the SQL engine. These are the basic operators that are used in almost any SQL query. 
The operators are stateless.

### PR DESCRIPTION
**Pull request by devozerov**
Implemented project and filter operators.

Limitations:
1. We have hard-coded batch size inside the `FilterExec` for now because we do not have a memory management yet. In the future we will switch to dynamic memory allocation.

Closes #16893

### CODE REVIEW

**Comment by taburet** [comment]
The build is failing with spotbugs errors.

**Comment by devozerov** [comment]
Fixed.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/Expression.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+package com.hazelcast.sql.impl.expression;
+
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+/**
+ * Defines expression contract for SQL.
+ *
+ * <p>Java serialization is needed for Jet.
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/expression/Expression.java:26]: It's no longer `Serializable` :)

Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/expression/Expression.java:26]: Fixed.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/182/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/182/Hazelcast-pr-builder-182.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CacheExpirationTest.test_whenEntryIsRemovedBackupIsCleaned:310 Expected to remove entry 0 but entry was not present. Expired entry count: 1
[INFO] 
[ERROR] Tests run: 27685, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by devozerov** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/sql/impl/exec/AbstractFilterExec.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.row.RowBatch;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Abstract filter executor that removes rows from the output based on a condition.
+ * <p>
+ * Currently the executor batches rows, and reports progress only when the batch is full or when EOS has been raeched.
```
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/exec/AbstractFilterExec.java]: Fixed.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/exec/FilterExec.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        super(id, upstream);
+
+        this.filter = filter;
+    }
+
+    @Override
+    protected boolean eval(Row row) {
+        Boolean res = filter.eval(row, ctx);
+
+        return res != null && res;
```
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/exec/FilterExec.java:39]: SQL may have three outcomes of predicate evaluation: `true`, `false`, or `null`. An example is `value OR NULL => NULL`

`null` results should be filtered out.

**Pull request merged**